patroni/postgresql.py

```diff
@@ -698,31 +690,28 @@ def start(self, block_callbacks=False):
 
         self.set_state('starting')
 
-        env = {'PATH': os.environ.get('PATH')}
-        # pg_ctl will write a FATAL if the username is incorrect. exporting PGUSER if necessary
-        if 'username' in self._superuser and self._superuser['username'] != os.environ.get('USER'):
-            env['PGUSER'] = self._superuser['username']
-
         self._write_postgresql_conf()
         self.resolve_connection_addresses()
 
-        options = ' '.join("--{0}='{1}'".format(p, self._server_parameters[p]) for p, v in self.CMDLINE_OPTIONS.items()
+        options = ' '.join('--{0}="{1}"'.format(p, self._server_parameters[p]) for p, v in self.CMDLINE_OPTIONS.items()
                            if self._major_version >= v[2])
 
+        cmd = 'exec "{0}" -D "{1}" {2} < "/dev/null" 2>&1'.format(self._pgcommand('postgres'), self._data_dir, options)
+
         start_initiated = time.time()
-        ret = self.pg_ctl('start', '-o', options, env=env, preexec_fn=os.setsid)
-        start_completed = time.time()
-        if ret:
-            # We want postmaster to open ports before we continue
-            try:
-                start_timeout = float(self.config.get('pg_ctl_timeout', 60))
-            except ValueError:
-                start_timeout = 60
-            ret = self.wait_for_port_open(start_timeout, start_initiated, start_completed)
+        pid = os.fork()
+        if pid == 0:
+            os.setsid()
+            os.execle("/bin/sh", "/bin/sh", "-c", cmd, {'PATH': os.environ.get('PATH')})
```

Comment:
Not a Windows expert by far, but I'll hazard a guess that this doesn't work on Windows :) Do we care?

Before execing there is probably some cleanup stuff to do, I know that watchdog support wants that the watchdog device file handle is closed and not passed on to postgres. Probably good to close the rest too. Why did you decide to go through `/bin/sh` and `os.execle`? Using `subprocess.Popen` to execute postgres directly might be more portable and give us stuff like close_fds.